libraries/chain/db_update.cpp

```diff
@@ -590,14 +590,13 @@ void database::clear_expired_htlcs()
          && htlc_idx.begin()->conditions.time_lock.expiration <= head_block_time() )
    {
       const htlc_object& obj = *htlc_idx.begin();
-      adjust_balance( obj.transfer.from, asset(obj.transfer.amount, obj.transfer.asset_id) );
-      // virtual op
-      htlc_refund_operation vop( obj.id, obj.transfer.from );
-      vop.htlc_id = htlc_idx.begin()->id;
+      const auto amount = asset(obj.transfer.amount, obj.transfer.asset_id);
+      adjust_balance( obj.transfer.from, amount );
+      // notify related parties
+      htlc_refund_operation vop( obj.id, obj.transfer.from, obj.transfer.to, amount,
+         obj.conditions.hash_lock.preimage_hash, obj.conditions.hash_lock.preimage_size );
```

Comment:
Simply adding fields to the vop does not automatically notify new parties unless also added code to `get_impacted_account_visitor::operator()( const htlc_refund_operation& op )` in `db_notify.cpp`. However, if not implemented carefully, account history (`2.9.x`) numbering would be affected which would break 3rd client apps who rely on consistent numbering. The careful approach is similar to what's done in #1629.

Reply:
I thought I added them to _impacted. Will do. I will also add the boolean based on HF date to prevent account history numbering issues.

Reply:
Upon further research,
1. I didn't write this, so that explains why I didn't add the necessary code :-)
2. [the original ticket](issues/1887) mentions it not needing a hard fork since the virtual op does not involve consensus. Therefore we have an issue of when should we turn on this feature. My suggestion would be to make it "hardfork-like" and wrap it with hardfork protections tied to BSIP 64.

Your thoughts please @abitmore

Reply:
Thanks for the link to the original ticket. Just realized that what I wrote in the reviews are actually new features. So, I think we will only work on this if we have enough time.

Yes, my thought was to make it hardfork-like. For implementation, I'd like to refactor Nathan's code, pass in a time rather than several booleans, write different logic for different operations.